Feature/active label

JobsIsMyHomeboy opened the pull request:
This adds a new subclass of UILabel called ActiveLabel. It can be used to show activity for a label while maintaining some spacing on the screen as well as visual indicator that something is happening. The use of nil on the label is what turns the activity indicator on and off.

JobsIsMyHomeboy committed:
- Added ActiveLabel Subclass.
- Modified podspec for ActiveLabel subspec.
- Increased version to 1.3.6
- Updated readme with details for ActiveLabel.
- Added ActiveLabel to Examples
- Added Unit Tests for ActiveLabel.

JobsIsMyHomeboy committed:
Updated example images for better quality.

JobsIsMyHomeboy committed:
Revert "Updated example images for better quality."

This reverts commit 03639e52a6937af9b829830e619349ee5f1055a4.

JobsIsMyHomeboy committed:
Updated Readme.

JobsIsMyHomeboy committed:
Updated Readme.

wmcginty commented on Examples/UtiliKit-iOSExample/ViewController/ActiveLabelViewController.swift:
```diff
@@ -0,0 +1,18 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        activeLabels.forEach({ $0.text = nil })
```
We should use trailing closure syntax here.

wmcginty commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+//
+
+import UIKit
+import GLKit
```
Wait, GLKit?

JobsIsMyHomeboy replied:
Oops. That was leftover from using some of the math functions. I'll remove it.

wmcginty commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+import GLKit
+
+struct ActiveLabelConfiguration {
+    var estimatedNumberOfLines: UInt
```
Do these `var` actually change frequently? Or can they be `let` ?

JobsIsMyHomeboy replied:
They don't change frequently, but they can be changed. I'm not bound to them being vars though, I just think it's easier if you only want to change a few of these based off of the default configuration.

wmcginty commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+    private static let locationKeyPath = "locations"
+    
+    private var isDisplayingInStoryboard: Bool = false
+    private(set) var isLoading: Bool = false {
```
Can we make the loading vs non-loading more clear? Maybe using an enum instead?

```swift
enum State {
case loading
case text(String)
}

JobsIsMyHomeboy replied:
I've replaced isLoading with a State enum. Let me know what you think.

tylermilner commented on README.md:
````diff
@@ -181,7 +181,48 @@ func containerViewController(_ container: ContainerViewController, animationCont
+``` 
+
+#### ActiveLabel
+ActiveLabel is a UILabel subclass that adds horizontal activity indicators to your label while the text value is set to nil. You can customize this view quite a bit in code or in Interface Builder to fit your specific needs. The purpose of this subclass is to have a visual indication at the label level while you are loading data into labels.
````
We might want to tweak the wording of "value" to "property" and also add some backticks for better syntax highlighting.
```
`ActiveLabel` is a `UILabel` subclass that adds horizontal activity indicators to your label while its `text` property is set to `nil`. You can customize this view quite a bit in code or in Interface Builder to fit your specific needs. The purpose of this subclass is to have a visual indication at the label level while you are loading data into labels.
```

tylermilner commented on Examples/UtiliKit-iOSExample/ViewController/ActiveLabelViewController.swift:
```diff
@@ -0,0 +1,18 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        activeLabels.forEach({ $0.text = nil })
```
Just some initial thoughts, but setting to `nil` might not be the most flexible. I'm thinking of a scenario like this:

1. Initial load (`text` property set to `nil` so the gradient animation is happening).
2. We get our data and set `text` to whatever value it needs to be (gradient animation disappears).
3. We need to clear out the text (set it back to `nil`), but we don't want the gradient animation to happen this time. It seems like this isn't currently possible (though I'm not sure if this is really a good real-world example).

JobsIsMyHomeboy replied:
I also thought about having a bool that would allow you to enable/disable the automatic animation when the label is set to nil. My worry was just that it might over complicate the use of this. I will go ahead and explore that though since not having that ability does limit the enduser when using this class.

tylermilner replied:
Yeah, I can see how it would make things more complicated. I don't necessarily think we need to solve it right now though, since I just made up the scenario in my head. 😄

tylermilner commented on README.md:
````diff
@@ -181,7 +181,48 @@ func containerViewController(_ container: ContainerViewController, animationCont
+label.finalLineTrailingInset = 100
+label.configurationChanged()
+```
+![](ActiveLabelEdited01.gif)
````
These images should probably live inside some sort of `docs/images` folder rather than the repo root.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+
+/// Used to show loading progress in a UILabel through an animated gradient.
+///
+/// Setting text to nil will cause the loading indicator to show, non nil will remove the loading indicators.
```
Let's add some backticks around `nil`.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+///
+/// Setting text to nil will cause the loading indicator to show, non nil will remove the loading indicators.
+/// Use lastLineTrailingInset and lastLineLength to control how the last line is displayed. If only 1 line then
+/// it is considered the last line. lastLineTrailingInset takes precendence over lastLineLength.
```
Same here - backticks around `lastLineTrailingInset`, `lastLineLength`, and `lastLineTrailingInset` so that we get the syntax highlighting when alt-clicking in Xcode.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+    /// Line Length in points for the last activity line. If finalLineTralingInset is set to greater than 0 this value is not used. Default is 0.
+    @IBInspectable public var finalLineLength: CGFloat = 0
+    /// This color is the darkest area of the line seen during activity animation. Default is (233,231,237) Gray.
+    @IBInspectable public var loadingViewColor: UIColor = UIColor(red: 233.0/255.0, green: 231.0/255.0, blue: 237.0/255.0, alpha: 1.0)
```
We have this gray color defined here as well as in `ActiveLabelConfiguration.default`. Can we move the definition into a single spot to be more DRY?

JobsIsMyHomeboy replied:
Are you thinking just a private static let like locationKeyPath is?

tylermilner replied:
Yeah, just a `static` property somewhere.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+import UIKit
+import GLKit
+
+struct ActiveLabelConfiguration {
```
Very minor, but we might consider moving this inside of `ActiveLabel` to get better namespacing (e.g. `ActiveLabel.Config`).

JobsIsMyHomeboy replied:
Great Idea. I'll get that moved inside the class.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+    }
+    
+    /**
+     When text is set to nil the loading views will display. When text is non-nill the loading views will be hidden.
```
Small typo: "non-nill" --> "non-nil".

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        
+        defer {
```
Do we need `defer` (since there's no other logic inside these initializers)?

JobsIsMyHomeboy replied:
I don't think we do. I'll pull it and test to make sure.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+     You must call this function to reset the loading views after you have changed a configuration value.
+     */
+    func configurationChanged() {
+        loadingViews.forEach({ $0.removeFromSuperview() })
```
Are the parenthesis necessary here? I think we can just use trailing closure syntax, right?
```swift
loadingViews.forEach { $0.removeFromSuperview() }
```

JobsIsMyHomeboy replied:
Not necessary, just a habit of mine. :)

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+        loadingViews.forEach({ $0.isHidden = true })
+    }
+    
+    func configureLoadingViews() {
```
Since this is a long method, it might be good to break it up into a couple of methods (view creation vs constraint configuration) and/or add some comments (especially around the index checks towards the bottom).

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,285 @@
+        if showLoadingViewsInStoryboard && isDisplayingInStoryboard {
+            maskLayer.locations = [NSNumber(value: 0.4), NSNumber(value: 0.5), NSNumber(value: 0.6)]
+        } else {
+            maskLayer.locations = [NSNumber(value: 0.0), NSNumber(value: 0.1), NSNumber(value: 0.2)]
```
What's the reasoning behind setting `maskLayer.locations` differently for storyboard rendering vs non-storyboard?

JobsIsMyHomeboy replied:
Changing the locations allows you to see the gradient in the storyboard since normal use the graduated part is clipped by the view. I wanted the developer to be able to see what the gradient would look like as they were designing it.

tylermilner replied:
👍 Let's just add a comment that explains that reasoning.

tylermilner commented on Tests/ActiveLabelTests.swift:
```diff
@@ -0,0 +1,125 @@
+import UIKit
+@testable import UtiliKit
+
+class ActiveLabelTests: XCTestCase {
```
Not sure how feasible this is, but I'd love to see some snapshot tests for this library (for the various configurations - like in the example gifs).

JobsIsMyHomeboy replied:
Which library would you like me to use for those?

tylermilner replied:
The `iOSSnapshotTestCase` should be fine. I know there's a newer snapshotting library out there too, which I'm not opposed to using. Up to you!

wmcginty replied:
The point free snapshot testing library is glorious.

JobsIsMyHomeboy replied:
Added Snapshot tests using the PointFree SnapshotTesting SDK.

tylermilner commented on UtiliKit.podspec:
```diff
@@ -8,7 +8,7 @@
 Pod::Spec.new do |s|
 s.name             = 'UtiliKit'
-s.version          = '1.3.5'
+s.version          = '1.3.6'
```
Let's not bump the version just yet. We'll do a separate release after this is merged into `master`.

JobsIsMyHomeboy replied:
OK, I'll change that back.

tylermilner reviewed (changes requested):
One question I have is how does this behave when using dynamic type? Do the gradient animations scale accordingly?

JobsIsMyHomeboy committed:
- Moved readme images into docs/images folder.
- Added some backticks to the ActiveLabel documentation.

JobsIsMyHomeboy committed:
- Moved ActiveLabelConfiguration inside ActiveLabel class.
- Added some backticks to the documentation.
- Created loadingGray static class constant to be used for the default color.
- Removed use of defer in initializers since it isn't needed.
- Moved constraint setup into its own function which is called when configuring the loading views.
- Cleaned up code.

JobsIsMyHomeboy committed:
- Changed version back to 1.3.5.

JobsIsMyHomeboy committed:
- Using trailing closure syntax now.

JobsIsMyHomeboy committed:
- Added SnapshotTesting framework from pointfree using Carthage.
- Modified UnitTests and Added specific Snapshot Tests.

JobsIsMyHomeboy committed:
Merge branch 'master' into feature/ActiveLabel

JobsIsMyHomeboy committed:
Added State enum to ActiveLabel class.
Replaced isLoading Bool to now use the State enum.
Tweaked documentation.

JobsIsMyHomeboy committed:
Merge branch 'feature/ActiveLabel' of github.com:JobsIsMyHomeboy/iOS-UtiliKit into feature/ActiveLabel

JobsIsMyHomeboy committed:
Added documentation to ActiveLabel class to point out that the loading indicators do not scale with Dynamic Type.

JobsIsMyHomeboy commented:
@tylermilner  Due to some weirdness with trying to dynamically size the loading views based on the Content Size Category change I don't feel like supporting it at this time is worthwhile. I have some ideas on how to handle it, but I don't think they would be worth introducing in a version 1 of this class. I've added a Note to the class documentation calling out that the loading views don't support resizing via dynamic type.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+    static let loadingGray: UIColor = UIColor(red: 233.0/255.0, green: 231.0/255.0, blue: 237.0/255.0, alpha: 1.0)
+    
+    /// Struct used to represent how an `ActiveLabel` should be displayed. Should be passed into the `ActiveLabel` convenience initializer.
+    struct ActiveLabelConfiguration {
```
We could name this `Configuration` since its namespaced under `ActiveLabel` (e.g. `ActiveLabel.Configuration` vs `ActiveLabel.ActiveLabelConfiguration`).

JobsIsMyHomeboy replied:
Fixed.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+    }
+    
+    /// The number of activity lines to display. Default is 1.
+    @IBInspectable public var estimatedNumberOfLines: UInt = 1
```
I don't think it will cause any big issues with maintenance down the road, but we could reference `ActiveLabelConfiguration.default` to get these default values rather than redefining them.
```swift
@IBInspectable public var estimatedNumberOfLines: UInt = ActiveLabelConfiguration.default.estimatedNumberOfLines
```

JobsIsMyHomeboy replied:
Fixed.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+    private static let locationKeyPath = "locations"
+    
+    /// When true the gradient will show centered so that it can be adjusted easily in IB
+    private var isDisplayingInStoryboard: Bool = false
```
Is there anyway to merge `isSnapshotTesting` and `isDisplayingInStoryboard`? It seems like they both center the gradient so maybe we could just have one `displayStaticGradient` or `isGradientCentered` property?

JobsIsMyHomeboy replied:
Removed isSnapshotTesting and isDisplayedInStoryboard var's in favor of a private var for isGradientCentered. It is now set to true in prepareForInterfaceBuilder if the text string in IB is non nil since you can't have a nil text value in IB. Also, created public function configureForSnapshotTest() which will also set the gradient to be centered for snapshots.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+    /// When true the gradient will show centered so that it can be adjusted easily in IB
+    private var isDisplayingInStoryboard: Bool = false
+    /// Public read-only representation of the labels state. Either `loading` or `text(String)`.
+    private(set) var state: State = .text("") {
```
I'm not sure how I feel about using an empty string to represent the initial "state" of this label. Don't labels usually start with their `text` property set to `nil`? Seems like there will be an opportunity for `state` and `text` to become out of sync. Also, there is a very subtle (but rarely meaningful) difference in meaning between an empty string and `nil`.

JobsIsMyHomeboy replied:
I've changed State to now accept an optional String so that it lines up with the standard text value of UILabel. If the labels text is nil the state is set to loading, but a non optional value like an empty string is now treated as not loading, but an empty label. There is a slight change in prepareForInterface builder that treats an empty string as a null value, but that is only for displaying the centered gradient in IB and does not display loading at runtime.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+    /**
+     You must call this function to reset the loading views after you have changed a configuration value.
+     */
+    func configurationChanged() {
```
Does this method need to be publicly accessible? I think most views in UIKit auto-update after you change one of their properties (e.g. `backgroundColor` on a `UIView` doesn't require you to manually call some sort of `configurationChanged()` function on the `UIView` instance). Maybe we should add `didSet`s (that call this function) to all of the `IBInspectable`s above?

JobsIsMyHomeboy replied:
Added didSet's to all configuration properties and moved configuration change to the private.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+
+private extension ActiveLabel {
+    func textDidUpdate() {
+        state = text == nil ? .loading : .text(text ?? "")
```
Very minor, but a "Swiftier" way to do this might be to add a custom initializer on `ActiveLabel.State`. Also, I think we might be able to accomplish it without using the nil-coalescing operator.
```swift
enum State: Equatable {
    case loading
    case text(String)

    init(text: String?) {
        guard let text = text else { self = .loading }
        self = .text(text)
    }
}
```

JobsIsMyHomeboy replied:
Fixed.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+            // Constrain the first view to the top of the label.
+            view.topAnchor.constraint(equalTo: topAnchor).isActive = true
+        } else {
+            // If not the first view constrain the top to the above views bottom with the loadingLineVerticalSpacing.
```
Very minor typo (missing comma and apostrophe).
```swift
// If not the first view, constrain the top to the above view's bottom with the loadingLineVerticalSpacing.
```

JobsIsMyHomeboy replied:
Fixed.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+        }
+        
+        if index + 1 == estimatedNumberOfLines {
+            // If this is the last view then constraint it's bottom to the label
```
Very minor typo (extraneous apostrophe).
```swift
// If this is the last view, then constraint its bottom to the label
```

JobsIsMyHomeboy replied:
Fixed.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -0,0 +1,324 @@
+        
+        maskLayer.colors = [lightColor.cgColor, darkColor.cgColor, lightColor.cgColor]
+        // If we are viewing in InterfaceBuilder we want to shift the gradient to the center so that it can be seen at design time otherwise set it to the far left.
+        if (showLoadingViewsInStoryboard && isDisplayingInStoryboard) || isSnapshotTesting {
```
Similar to above comment about `isDisplayingInStoryboard` and `isSnapshotTesting` - maybe `showLoadingViewsInStoryboard` could be combined into that as well?

JobsIsMyHomeboy replied:
Changed to use new isGradientCentered private var.

JobsIsMyHomeboy committed:
Renamed ActiveLabelConfiguration to Configuration since it is namespaced under ActiveLabel.
Changed ActiveLabel configuration properties to used the Configuration.default values.
Removed showLoadingViewsInStoryboard and isSnapshotTesting in favor of a private isGradientCentered bool to signal centering the gradient.
Added public function configureForSnapshotTesting that sets isGradientCentered to true in order to keep that variable private.

JobsIsMyHomeboy committed:
Added State initializer that takes an optional string.
Moved configurationChanged() to private and added didSet to all configuration properties that now call configurationChanged()
Updated default state value to .text(nil)
Added configuredForSnapshotTest public function in order to center the gradient for snapshot tests.
Fixed textDidUpdate to use new State initializer.
Corrected some comments.
Fixed tests to work properly with latest changes.
Updated ReadMe to reflect changes.
Added line to ReadMe about configuring for snapshot tests.

JobsIsMyHomeboy commented:
I just pushed an update that addresses the open issues for the PR.

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -54,7 +54,16 @@ class ActiveLabel: UILabel {
+                return
+            }
+            
+            self = .text(text)
```
No need to change this, but it could also be done in one line with `flatMap`. 😄 
```swift
self = text.flatMap { .text($0) } ?? .loading
```

tylermilner commented on Sources/UtiliKit/ActiveLabel/ActiveLabel.swift:
```diff
@@ -70,34 +79,69 @@ class ActiveLabel: UILabel {
+    }
+//    /// Shows the loading views in the Storyboard by default since text of a label can't be nil in a Storybaord. Default is true.
+//    @IBInspectable private var showLoadingViewsInStoryboard: Bool = true
+//
```
Looks like there's some properties and comments that are commented out.

JobsIsMyHomeboy replied:
Oops. I'll get those removed.

tylermilner reviewed (commented):
@JobsIsMyHomeboy I think this should be good to go after you add a `CHANGELOG` entry!

JobsIsMyHomeboy committed:
Removed old commented out code.

JobsIsMyHomeboy committed:
Updated travis.yml to add support for Carthage since snapshot tests are now done with a library installed via Carthage.

JobsIsMyHomeboy committed:
Updated Changelog to show ActiveLabel addition.

JobsIsMyHomeboy committed:
Changed OS in travis to 12.2.

JobsIsMyHomeboy committed:
Fixed ActiveLabel subspec.

tylermilner commented:
@JobsIsMyHomeboy It looks like [Codebeat is complaining](https://codebeat.co/projects/github-com-bottlerocketstudios-ios-utilikit-master/pull_requests/224120) about a few issues regarding `ActiveLabel`. Can you take a look and see what you can easily fix?

JobsIsMyHomeboy commented:
@wmcginty @tylermilner I don't have write access so I can't merge. :(

wmcginty commented:
Is there anything we can do to fix any/some of those Codebeat issues @JobsIsMyHomeboy ?

JobsIsMyHomeboy committed:
Fixed Codebeat error for too many instance variables on ActiveLabel.Configuration by adding a secondary struct for the LoadingView portion of the variables.
Updated ReadMe and fixed tests with configuration changes.

JobsIsMyHomeboy committed:
Refactored out animation part creation into its own method. This should resolve the function too long warning.
